Implement iterative average structure (Issue#2039)

leonwehrhan opened the pull request:
Fixes #2039 

Changes made in this Pull Request:
 - Implemented average structures with iterative algorithm from DOI 10.1021/acs.jpcb.7b11988. (Issue #2039)


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).


<!-- readthedocs-preview mdanalysis start -->
----
📚 Documentation preview 📚: https://mdanalysis--4524.org.readthedocs.build/en/4524/

<!-- readthedocs-preview mdanalysis end -->

pep8speaks commented:
Hello @leonwehrhan! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2024-03-29 18:38:13 UTC

yuxuanzhuang commented:
Hello @leonwehrhan! Welcome to MDAnalysis! Please add #2039 directly after "Fixes" so the issue will be successfully closed after this PR is merged.

Feel free to ping us when you feel it's ready for review :)

leonwehrhan committed:
Implement iterative average structure (Issue#2039)

leonwehrhan commented:
Hello @yuxuanzhuang, I think the PR is ready for a review. 

Could you maybe give some guidance on how to resolve the failed automatic checks? I am not sure how to tackle the issue with the mypy linter, as my local mypy does not see any issues with the files I have touched. Also, my tests should cover every line of new code (codecov/patch 100%) so I do not know why the codecov/project test failed.

Thank you very much!

yuxuanzhuang commented:
Hi, there are a few aspects up for discussion (also pinging @lilyminium, since you added the `AverageStructure`):

1) Should it be a method or an Analysis class? It also makes sense to include it as an option to run `AverageStructure` (e.g., `iterative=True`). For now, I think it's acceptable to leave it as is.

2) In scenarios where you have a large system and want to obtain the average structure of, for example, a protein, the current implementation does not offer the option to do subselections, which `AverageStructure` can.

3) The code you provided doesn't actually accomplish the task. If you check for verbosity, you can see that the RMSD is not changing at all. This is because `ref.positions = res.positions` does not actually modify the position of atoms in the reference universe. What you can do is use `ref = res.universe`, which would represent the average structure from the last round of iteration. Please double-check the output of your code.

4) The test also fails to achieve its goal (except for checking for regression). One problem is that `u1 = mda.Merge(u.select_atoms("bynum 1:10"))` turns the universe into a single-frame trajectory (check `u1.trajectory`).

Don't worry about the failing of linter and main_tests (ubuntu-latest, 3.12, true, true), they are unrelated issues.

leonwehrhan committed:
Update after review: Implement iterative average structure (Issue#2039)

leonwehrhan commented:
Hi @yuxuanzhuang,

thank you very much for the review! I have updated the code and adressed your concerns as follows:

1. While I also think an implementation as part of the `AverageStructure` class makes sense, I have left the method as it is. I could open another issue and work on it once this PR is merged?
2. I have implemented the optional `select` keyword. As the reference structure is updated throughout the iterative calculation, the selection for the reference is set in the beginning, by transforming the reference universe into a one-frame universe of the selected atoms of the current frame of the reference. The selection for the mobile universe is passed to the `AverageStructure` class in each iteration.
3. I have implented the code you suggested. I checked the output using `verbose=True` to see the RMSD is changing. I ran `iterative_average` on `u=mda.Universe(PSF, DCD)` and visually checked that the resulting structure looked reasonable.
4. Given the implementation of the `select` keyword, it is now not necessary to use an atom-sliced universe as input for the iterative workflow, so the original universe `u` is now returned and processed in the tests. I included some regression tests and some tests checking that the output of the first 10 atoms is the same that I have visually checked.

Other things I have changed:
- The `weights` keyword could not be actually used before, because calling `rms.rmsd` would produce an error. This is fixed by calling `get_weights` and is included in the tests.
- The default `eps=1e-8` was not really useful to get to a converged structure (even after 1000 iterations), so I set it to `1e-6`.
- Updated documentation.

yuxuanzhuang commented on package/CHANGELOG:
```diff
@@ -62,6 +62,12 @@ Deprecations
    (PR #4403)
 
 
+03/22/24 leonwehrhan
```
move it to the block for release 2.8.0. Add your name to the end of the other authors (of that release)

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+    ----------
+    mobile : mda.Universe
+        Universe containing trajectory to be fitted to reference.
+    ref : mda.Universe
```
make it optional (if not set, then use mobile as reference). change the name to `reference` to match AverageStructure.

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+    -------
+    res : AverageStructure
+        AverageStructure result from the last iteration.
+
```
add a short code snippet as an example

yuxuanzhuang replied:
including how to retrieve the final Universe from the output

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+
+    Returns
+    -------
+    res : AverageStructure
```
res is a confusing name, I would recommend av (or avg_struc)

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+    res : AverageStructure
+        AverageStructure result from the last iteration.
+
+    """
```
add `versionadded` remark (see other code as examples)

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+    The optimal reference is defined as average structure of a
+    trajectory, with the optimal reference used as input.
+    This function computes the optimal reference by using a starting
+    reference for the average structure, which is used to calculate the
```
```suggestion
    reference for the average structure, which is used as the reference to calculate the
```

yuxuanzhuang commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -332,6 +334,78 @@ def process_selection(select):
+    This function computes the optimal reference by using a starting
+    reference for the average structure, which is used to calculate the
+    average structure again. This is repeated until the reference
+    structure has converged.
```
add citation inside package/doc/sphinx/source/references.bib.

add :footcite:p:`xxxxx` here.

yuxuanzhuang reviewed (changes requested):
Looks great! Just a few comments on the documentation side.

And please fix pep8 issue reported above.

leonwehrhan committed:
Update 2 after review: Implement iterative average structure (Issue#2039)

yuxuanzhuang commented:
I think this function should actually be put inside the `analysis.align` module since what it does is alignment. Could you move the code (and tests) there?

The other issue is the documentation is not showing up on the webpage (there's a preview you can find above https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/rms.html). You need to add explicitly `.. autofunction:: xxx` to the documentation at the beginning of the file.

Additionally, please leave the review unresolved so that the reviewer can determine whether the issue has indeed been addressed.

leonwehrhan committed:
Update 3 after review: Implement iterative average structure (Issue#2039)

leonwehrhan commented:
I have moved the function to `analysis.align` and added the remark to the documentation in the beginning of the file. The documentation of `iterative_average` now shows up on https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/align.html.

orbeckst commented:
@yuxuanzhuang I added you as the person responsible for the PR.

@leonwehrhan I un-assigned you because we use the "Assignee" field to indicate who from the developer team manages review and merge.

yuxuanzhuang reviewed (approved):
lgtm!

leonwehrhan committed:
Merge branch 'develop' into develop

orbeckst commented on package/CHANGELOG:
```diff
@@ -55,6 +55,8 @@ Changes
    equivalent assert_allclose(...  rtol=0, atol=1.5e-{N}) (issue modernize
    testing code #3743, PR Replaced numpy.testing.assert_almost_equal to 
    numpy.testing.assert_allclose #4438)
+ * Implement average structures with iterative algorithm from 
```
This is an Enhancement. Please move into the appropriate section.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+        av = rms.iterative_average(u, u, verbose=True)
+
+        averaged_universe = av.results.universe
+
```
Does this render correctly? Check!

If necessary, insert blank line.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
     return old_rmsd, new_rmsd
 
 
+def iterative_average(
```
add a duecredit decorator for the paper , see eg https://github.com/MDAnalysis/mdanalysis/blob/f5335b4db7f6f746f163e3a617ae6658d30d2fb9/package/MDAnalysis/analysis/bat.py#L257-L259 (look at code in the code base for how we use [duecredit](https://docs.mdanalysis.org/stable/documentation_pages/references.html#citations-using-duecredit).

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+    mobile, reference=None, select='all', weights=None, niter=100,
+    eps=1e-6, verbose=False, **kwargs
+):
+    """Calculate an optimal reference that is also the average structure
```
Follow the numpy doc string standard (see https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation ) and start with a single line summary:
```reST
Iteratively calculate an optimal reference that is also the average structure after an RMSD alignment.

The optimal reference is defined as average structure of a trajectory, with ...
```
and include "Iterative" there so that the function name makes more sense.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+    Example
+    -------
+
+    ::
```
Add a sentence saying in words what the example does.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+        u = mda.Universe(PSF, DCD)
+        av = rms.iterative_average(u, u, verbose=True)
+
+        averaged_universe = av.results.universe
```
Will this code be executed when you run the doc tests? — see #3925 

If yes, please make sure that it works.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+        ref = avg_struc.results.universe
+
+        if verbose:
+            print(
+                "i = {}, rmsd-change = {:.5f}, ave-rmsd = {:.5f}".format(
+                    i, drmsd, avg_struc.results.rmsd
+                )
+            )
```
We never use print. Use a logger at debug level.
```python
logger.debug(f"iterative_average(): i = {i}, rmsd-change = {drmsd:.5f}, ave-rmsd = {avg_struc.results.rmsd:.5f}")
```

orbeckst replied:
Alternatively, use a custom progress and show this information on the progress bar.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+                    i, drmsd, avg_struc.results.rmsd
+                )
+            )
+
```
Add a `logger.info()` message for the converged result that shows the final values.

orbeckst replied:
Add an `else:` clause to the `for` loop and fail with a `RuntimeError` if the procedure did not converge:
```python
else:
   errmsg = (f"iterative_average(): Did not converge in {niter} iterations to DRMSD < {tol}. "
           f"Final average RMSD = {avg_struc.results.rmsd:.5f}")
   logger.error(errmsg)
   raise RuntimeError(errmsg)
```
You should also add a test that triggers the exception (eg by running a test with low tol and niter).

When you have an iterative algorithm it's important to at least know when it fails.

orbeckst commented on package/doc/sphinx/source/references.bib:
```diff
@@ -770,3 +770,13 @@ @article{Kulke2022
+  journal = {The Journal of Physical Chemistry B},
+  volume = {122},
+  number = {21},
+  pages = {5630--5639}
```
add DOI

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -35,6 +35,7 @@
                                        ALIGN_UNBOUND, PDB_helix)
 from numpy.testing import (
     assert_equal,
+    assert_almost_equal,
```
just use  `assert_allclose` if possible, see  #3743

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -594,6 +595,63 @@ def test_sequence_alignment_deprecation(self, atomgroups):
+
+        res = align.iterative_average(mobile, reference, select='bynum 1:10',
+                                      niter=10, verbose=True)
+        assert_almost_equal(
```
just use assert_allclose (or is there a specific reason for   assert_almost_equal ??)

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -594,6 +595,63 @@ def test_sequence_alignment_deprecation(self, atomgroups):
+
+        res = align.iterative_average(mobile, reference, select='bynum 1:10',
+                                      niter=10, weights='mass')
+        assert_almost_equal(
```
assert_allclose

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -594,6 +595,63 @@ def test_sequence_alignment_deprecation(self, atomgroups):
             align.sequence_alignment(mobile, reference)
 
 
+class TestIterativeAverage(object):
```
Implement the convergence check (see above) and add a test for it.

orbeckst reviewed (changes requested):
Looking good but we need a convergence check and we need to handle text output via logging.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +542,97 @@ def alignto(mobile, reference, select=None, weights=None,
+    weights = get_weights(ref.atoms, weights)
+
+    drmsd = np.inf
+    for i in range(niter):
```
ProgressBar here would be nice!

leonwehrhan committed:
Update 4 after review: Implement iterative average structure (Issue#2039)

leonwehrhan committed:
Merge branch 'develop' of github.com:leonwehrhan/mdanalysis into develop

leonwehrhan commented:
@yuxuanzhuang @orbeckst I have made the requested changes.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -584,15 +593,17 @@ def iterative_average(
 
     Example
     -------
+    `iterative_average` can be used to obtain a `MDAnalysis.Universe` with
```
Use markup
```suggestion
    `iterative_average` can be used to obtain a :class:`MDAnalysis.Universe` with
```

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -607,15 +606,17 @@ def reference(self):
         res = align.iterative_average(mobile, reference, select='bynum 1:10',
                                       weights=np.ones(10))
 
+        with pytest.raises(RuntimeError):
+            _ = align.iterative_average(mobile, reference, niter=1, eps=0)
+
```
Make this a separate test method `test_iterative_average_convergence_failure()`.

We want tests to be fine grained and testing one thing.

orbeckst commented on testsuite/MDAnalysisTests/analysis/test_align.py:
```diff
@@ -607,15 +606,17 @@ def reference(self):
+        res = align.iterative_average(mobile, niter=10)
         res = align.iterative_average(mobile, reference, eps=1e-5)
         res = align.iterative_average(mobile, reference, select='bynum 1:10',
                                       weights=np.ones(10))
```
These are just running through code paths and they do not assert anything. These are only marginally useful tests.

Please split `test_iterative_average` into multiple testing methods where each method tests one thing (eg default values, different eps, weights, ...) and where you 
1. run iterative_average() _once_
2. assert that the output is as expected

This will be more code but in this way we will see exactly if one specific part of the code fails.

orbeckst reviewed (changes requested):
Thanks for making the changes. 

My main comment is for the tests: please split your tests so that you always just test one thing as this will help tremendously in debugging if anything ever fails. Please see the comments inline.

leonwehrhan committed:
added separate tests

leonwehrhan commented:
@orbeckst I have changed the tests, so that every testing method only contains one call of `iterative average` and changed the documentation as you suggested.

leonwehrhan committed:
Merge branch 'develop' into develop

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +544,120 @@ def alignto(mobile, reference, select=None, weights=None,
+    This function computes the optimal reference by using a starting
+    reference for the average structure, which is used as the reference
+    to calculate the average structure again. This is repeated until the
+    reference structure has converged. :footcite:p:`Linke2018`
```
I just checked https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/align.html#MDAnalysis.analysis.align.iterative_average and the reference  `[1]` does not resolve.  Please check how to use footnote citations, e.g., as used in https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/bat.py 

I think you need to remove the reference from references and instead add a 
```reST

References
----------

.. footbibliography::

```
at the end of the module docs.

orbeckst commented on package/MDAnalysis/analysis/align.py:
```diff
@@ -541,6 +544,120 @@ def alignto(mobile, reference, select=None, weights=None,
+        ref = avg_struc.results.universe
+
+        if verbose:
+            logger.debug(
```
add a test that runs through this block, i.e., with `verbose=True`. You don't have to check for logger output but we need to check that there are no errors in the f-string.

orbeckst replied:
This will also help to increase coverage.

orbeckst reviewed (changes requested):
Thanks for changing the tests. Just found two small things. Almost there!!

leonwehrhan committed:
test and doc update

leonwehrhan commented:
@orbeckst @yuxuanzhuang the requested changes are pushed, but it looks like auto-merge has been disabled by my commit because I do not have write access. Could one of you do the merge?

orbeckst commented:
Auto-merge just means that the merge will be performed when all requirements are met
- all reviews are approving
- all required status checks passing

It's just a short cut for maintainers to essentially "queue up" the merge.

Let me have a quick look.

orbeckst commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -166,7 +166,9 @@
 
+import MDAnalysis as mda
 import MDAnalysis.lib.qcprot as qcp
+import MDAnalysis.analysis.align as align
```
Why was this import added? Can we remove it again? This almost certainly creates a circular import.
```suggestion
```

orbeckst commented on package/MDAnalysis/analysis/rms.py:
```diff
@@ -166,7 +166,9 @@
 import logging
 import warnings
 
+import MDAnalysis as mda
```
Why was this import added? 
```suggestion
```

orbeckst reviewed (changes requested):
Please check the imports that I flagged.

In order to remove them you can simply use my suggestions in the GH interface.

leonwehrhan commented:
Thanks for the clarification! I thought I might have unintentionally halted the merge process :)

I will adress your changes now.

leonwehrhan committed:
Update package/MDAnalysis/analysis/rms.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

leonwehrhan committed:
Update package/MDAnalysis/analysis/rms.py

Co-authored-by: Oliver Beckstein <orbeckst@gmail.com>

leonwehrhan commented:
@orbeckst I have comitted the changes via your comment. The reason for these imports being there was that I originally implemented the function in `analysis.rms` instead of `analysis.align` and later moved the code (see comments above).

orbeckst commented:
ok. the other test failure is not due to your stuff, see #4540

orbeckst commented:
Thank you for your contribution! 🎉

leonwehrhan commented:
Thanks a lot for guiding me through the PR!

yuxuanzhuang commented:
Congrats! @leonwehrhan
